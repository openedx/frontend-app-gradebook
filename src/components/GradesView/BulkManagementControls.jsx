/* eslint-disable react/sort-comp, react/button-has-type */
import React from 'react';
import PropTypes from 'prop-types';
import { connect } from 'react-redux';

import { views } from 'data/constants/app';
import actions from 'data/actions';
import selectors from 'data/selectors';

import NetworkButton from 'components/NetworkButton';
import ImportGradesButton from './ImportGradesButton';

import messages from './BulkManagementControls.messages';

/**
 * <BulkManagementControls />
 * Provides download buttons for Bulk Management and Intervention reports, only if
 * showBulkManagement is set in redus.
 */
export class BulkManagementControls extends React.Component {
  constructor(props) {
    super(props);
    this.handleClickExportGrades = this.handleClickExportGrades.bind(this);
    this.handleViewActivityLog = this.handleViewActivityLog.bind(this);
  }

  handleClickExportGrades() {
    this.props.downloadBulkGradesReport();
    window.location.assign(this.props.gradeExportUrl);
  }

  handleViewActivityLog() {
    this.props.setView(views.bulkManagementHistory);
  }

  render() {
    return this.props.showBulkManagement && (
      <div className="d-flex">
        <NetworkButton
          label={messages.downloadGradesBtn}
          onClick={this.handleClickExportGrades}
        />
        <ImportGradesButton />
      </div>
    );
  }
}

BulkManagementControls.defaultProps = {
  showBulkManagement: false,
};

BulkManagementControls.propTypes = {
  // redux
  downloadBulkGradesReport: PropTypes.func.isRequired,
  gradeExportUrl: PropTypes.string.isRequired,
  showBulkManagement: PropTypes.bool,
  setView: PropTypes.func.isRequired,
};

export const mapStateToProps = (state) => ({
  gradeExportUrl: selectors.root.gradeExportUrl(state),
  showBulkManagement: selectors.root.showBulkManagement(state),
});

export const mapDispatchToProps = {
  downloadBulkGradesReport: actions.grades.downloadReport.bulkGrades,
  setView: actions.app.setView,
};

export default connect(mapStateToProps, mapDispatchToProps)(BulkManagementControls);
