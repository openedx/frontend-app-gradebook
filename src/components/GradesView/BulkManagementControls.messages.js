import { defineMessages } from '@edx/frontend-platform/i18n';

const messages = defineMessages({
  downloadGradesBtn: {
    id: 'gradebook.GradesView.BulkManagementControls.bulkManagementLabel',
    defaultMessage: 'Download Grades',
    description: 'Button text for bulk grades download control in GradesView',
  },
});

export default messages;
