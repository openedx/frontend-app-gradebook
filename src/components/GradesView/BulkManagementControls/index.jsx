/* eslint-disable react/sort-comp, react/button-has-type */
import React from 'react';

import NetworkButton from 'components/NetworkButton';
import ImportGradesButton from '../ImportGradesButton';

import useBulkManagementControlsData from './hooks';
import messages from './messages';

/**
 * <BulkManagementControls />
 * Provides download buttons for Bulk Management and Intervention reports, only if
 * showBulkManagement is set in redus.
 */
export const BulkManagementControls = () => {
  const {
    show,
    handleClickExportGrades,
  } = useBulkManagementControlsData();

  if (!show) { return null; }
  return (
    <div className="d-flex">
      <NetworkButton
        label={messages.downloadGradesBtn}
        onClick={handleClickExportGrades}
      />
      <ImportGradesButton />
    </div>
  );
};

export default BulkManagementControls;
