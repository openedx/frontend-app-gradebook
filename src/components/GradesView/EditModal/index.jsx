import React from 'react';

import {
  Button,
  Alert,
  ModalDialog,
  ActionRow,
} from '@edx/paragon';
import { useIntl } from '@edx/frontend-platform/i18n';

import OverrideTable from './OverrideTable';
import ModalHeaders from './ModalHeaders';
import useEditModalData from './hooks';
import messages from './messages';

/**
 * <EditModal />
 * Wrapper component for the modal that allows editing the grade for an individual
 * unit, for a given student.
 * Provides a StatusAlert with override fetch errors if any are found, an OverrideTable
 * (with appropriate headers) for managing the actual override, and a submit button for
 * adjusting the grade.
 * (also provides a close button that clears the modal state)
 */
export const EditModal = () => {
  const { formatMessage } = useIntl();
  const {
    onClose,
    error,
    handleAdjustedGradeClick,
    isOpen,
  } = useEditModalData();

  return (
    <ModalDialog
      title={formatMessage(messages.title)}
      isOpen={isOpen}
      onClose={onClose}
      size="xl"
      hasCloseButton
      isFullscreenOnMobile
    >
      <ModalDialog.Body>
        <div>
          <ModalHeaders />
          <Alert variant="danger" show={!!error} dismissible={false}>
            {error}
          </Alert>
          <OverrideTable />
          <div>{formatMessage(messages.visibility)}</div>
          <div>{formatMessage(messages.saveVisibility)}</div>
        </div>
      </ModalDialog.Body>

      <ModalDialog.Footer>
        <ActionRow>
          <ModalDialog.CloseButton variant="tertiary">
            {formatMessage(messages.closeText)}
          </ModalDialog.CloseButton>
          <Button variant="primary" onClick={handleAdjustedGradeClick}>
            {formatMessage(messages.saveGrade)}
          </Button>
        </ActionRow>
      </ModalDialog.Footer>
    </ModalDialog>
  );
};

export default EditModal;
