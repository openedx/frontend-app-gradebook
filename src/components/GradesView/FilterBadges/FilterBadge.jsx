import React from 'react';
import { connect } from 'react-redux';
import PropTypes from 'prop-types';

import { Button } from '@edx/paragon';
import { FormattedMessage } from '@edx/frontend-platform/i18n';

import selectors from 'data/selectors';

/**
 * FilterBadge
 * Base filter badge component, that displays a name and a close button.
 * If showValue is true, it will also display the included value.
 * @param {func} handleClose - close/dismiss filter event, taking a list of filternames
 *   to reset when the filter badge closes.
 * @param {string} filterName - api filter name (for redux connector)
 */
export const FilterBadge = ({
  config: {
    displayName,
    isDefault,
    hideValue,
    value,
    connectedFilters,
  },
  handleClose,
}) => !isDefault && (
  <div>
    <span className="badge badge-info">
      <span>
        <FormattedMessage {...displayName} />
      </span>
      <span>
        {!hideValue ? `: ${value}` : ''}
      </span>
      <Button
        className="btn-info"
        aria-label="close"
        onClick={handleClose(connectedFilters)}
      >
        <span aria-hidden="true">&times;</span>
      </Button>
    </span>
    <br />
  </div>
);

FilterBadge.propTypes = {
  handleClose: PropTypes.func.isRequired,
  // eslint-disable-next-line
  filterName: PropTypes.string.isRequired,
  // redux
  config: PropTypes.shape({
    connectedFilters: PropTypes.arrayOf(PropTypes.string),
    displayName: PropTypes.shape({
      defaultMessage: PropTypes.string,
    }).isRequired,
    isDefault: PropTypes.bool.isRequired,
    hideValue: PropTypes.bool,
    value: PropTypes.oneOfType([
      PropTypes.string,
      PropTypes.bool,
    ]),
  }).isRequired,
};

export const mapStateToProps = (state, ownProps) => ({
  config: selectors.root.filterBadgeConfig(state, ownProps.filterName),
});

export default connect(mapStateToProps)(FilterBadge);
