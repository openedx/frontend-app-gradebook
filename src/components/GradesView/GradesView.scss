.spinner-overlay {
    position: fixed;
    height: 100%;
    width: 100%;
    top: 0;
    left: 0;
    background-color: #999;
    opacity: 0.5;
    z-index: 99999;
    display:flex;
    align-items: flex-start;
    justify-content: center;
    padding: 200px;
}

.color-black {
    color: black;
}

.gradebook-content {
    // note that this width isn't well-abstracted from Drawer code.
    // if we need to change it we may need to dig into those styles as well
    width: 100vw;
    .search-help-text {
        margin-left: 20px;
    }
    h4 {
        font-weight: bold;
        margin-top: 0rem;
    }
    .import-grades-btn {
      margin-left: 20px;
    }
    .intervention-report-description {
      margin-right: 40px;
    }
    h4.step-message-1 {
      margin-top: 40px;
    }
}

.student-filters{
    .label{
        padding-top: 30px;
    }
}
.grade-history-header{
    float: left;
}

.grade-history-assignment{
    padding-right: 49px;
}
.grade-history-student{
    padding-right: 75px;
}

.grade-history-original-grade{
    padding-right: 25px;
}

.grade-history-current-grade{
    padding-right: 25px;
}
.gradebook-container {
    width: 100%;
    overflow-x: auto;
    height: 600px;
    overflow-y: auto;
    position: relative;
}

.form-group, .pgn__form-group {
  label {
    font-weight: bold;
  }
}

.filter-group {
  .grade-filter-inputs {
    .percent-group {
      display: inline-block;
      .form-group, .pgn__form-group {
        width: 115px;
        display: inline-block;
      }
      .input-percent-label {
          margin-top: 22px;
          margin-left: 5px;
          margin-right: 5px;
      }
    }
  }
  .grade-filter-action {
    text-align: right;
  }
}

.mb-85 {
    margin-bottom: 85px;
}

.modal-dialog {
    max-width: 1000px;
}

.wrap-text-in-cell {
    overflow-wrap: break-word;
    word-wrap: break-word;
}

select#ScoreView.form-control {
  padding-right: 26px;
}

[dir=rtl] #course-grade-tooltip .arrow {
    right: initial;
    left: 0;
    
    &:before {
        border-width: 0.4rem 0.4rem 0.4rem 0;
        border-right-color: $black;
    }
}
