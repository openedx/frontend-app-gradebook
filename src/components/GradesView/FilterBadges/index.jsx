/* eslint-disable import/no-named-as-default */
import React from 'react';
import PropTypes from 'prop-types';

import { badgeOrder } from 'data/constants/filters';

import FilterBadge from './FilterBadge';

/**
 * FilterBadges
 * Displays a FilterBadge for each filter type in the data model with their current values.
 * @param {func} handleClose - event taking a list of filternames to reset
 */
export const FilterBadges = ({ handleClose }) => (
  <div>
    {badgeOrder.map(filterName => (
      <FilterBadge key={filterName} {...{ handleClose, filterName }} />
    ))}
  </div>
);
FilterBadges.propTypes = {
  handleClose: PropTypes.func.isRequired,
};

export default FilterBadges;
