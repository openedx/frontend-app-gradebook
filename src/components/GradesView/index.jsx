/* eslint-disable react/sort-comp, react/button-has-type, import/no-named-as-default */
import React from 'react';
import PropTypes from 'prop-types';
import { connect } from 'react-redux';

import { FormattedMessage } from '@edx/frontend-platform/i18n';

import actions from 'data/actions';
import thunkActions from 'data/thunkActions';

import BulkManagementControls from './BulkManagementControls';
import EditModal from './EditModal';
import FilterBadges from './FilterBadges';
import FilteredUsersLabel from './FilteredUsersLabel';
import FilterMenuToggle from './FilterMenuToggle';
import GradebookTable from './GradebookTable';
import ImportSuccessToast from './ImportSuccessToast';
import InterventionsReport from './InterventionsReport';
import PageButtons from './PageButtons';
import ScoreViewInput from './ScoreViewInput';
import SearchControls from './SearchControls';
import SpinnerIcon from './SpinnerIcon';
import StatusAlerts from './StatusAlerts';
import messages from './messages';

export class GradesView extends React.Component {
  constructor(props) {
    super(props);
    this.handleFilterBadgeClose = this.handleFilterBadgeClose.bind(this);
  }

  handleFilterBadgeClose(filterNames) {
    return () => {
      this.props.resetFilters(filterNames);
      this.props.updateQueryParams(filterNames.reduce(
        (obj, filterName) => ({ ...obj, [filterName]: false }),
        {},
      ));
      this.props.fetchGrades();
    };
  }

  render() {
    return (
      <>
        <SpinnerIcon />

        <InterventionsReport />
        <h4 className="step-message-1">
          <FormattedMessage {...messages.filterStepHeading} />
        </h4>

        <div className="d-flex justify-content-between">
          <FilterMenuToggle />
          <SearchControls />
        </div>

        <FilterBadges handleClose={this.handleFilterBadgeClose} />
        <StatusAlerts />

        <h4><FormattedMessage {...messages.gradebookStepHeading} /></h4>

        <div className="d-flex justify-content-between align-items-center mb-2">
          <ScoreViewInput />
          <BulkManagementControls />
        </div>

        <FilteredUsersLabel />

        <GradebookTable />

        <PageButtons />
        <p>* <FormattedMessage {...messages.mastersHint} /></p>
        <EditModal />

        <ImportSuccessToast />
      </>
    );
  }
}

GradesView.defaultProps = {};

GradesView.propTypes = {
  updateQueryParams: PropTypes.func.isRequired,

  // redux
  fetchGrades: PropTypes.func.isRequired,
  resetFilters: PropTypes.func.isRequired,
};

export const mapStateToProps = () => ({});

export const mapDispatchToProps = {
  fetchGrades: thunkActions.grades.fetchGrades,
  resetFilters: actions.filters.reset,
};

export default connect(mapStateToProps, mapDispatchToProps)(GradesView);
