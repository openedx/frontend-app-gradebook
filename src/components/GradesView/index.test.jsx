import React from 'react';
import { shallow } from 'enzyme';

import FilterBadges from './FilterBadges';

import useGradesViewData from './hooks';
import GradesView from '.';

jest.mock('./BulkManagementControls', () => 'BulkManagementControls');
jest.mock('./EditModal', () => 'EditModal');
jest.mock('./FilterBadges', () => 'FilterBadges');
jest.mock('./FilteredUsersLabel', () => 'FilteredUsersLabel');
jest.mock('./FilterMenuToggle', () => 'FilterMenuToggle');
jest.mock('./GradebookTable', () => 'GradebookTable');
jest.mock('./ImportSuccessToast', () => 'ImportSuccessToast');
jest.mock('./InterventionsReport', () => 'InterventionsReport');
jest.mock('./PageButtons', () => 'PageButtons');
jest.mock('./ScoreViewInput', () => 'ScoreViewInput');
jest.mock('./SearchControls', () => 'SearchControls');
jest.mock('./SpinnerIcon', () => 'SpinnerIcon');
jest.mock('./StatusAlerts', () => 'StatusAlerts');
jest.mock('./hooks', () => jest.fn());

const hookProps = {
  stepHeadings: {
    filter: 'filter-step-heading',
    gradebook: 'gradebook-step-heading',
  },
  handleFilterBadgeClose: jest.fn().mockName('hooks.handleFilterBadgeClose'),
  mastersHint: 'test-masters-hint',
};
useGradesViewData.mockReturnValue(hookProps);

const updateQueryParams = jest.fn().mockName('props.updateQueryParams');

let el;
describe('GradesView component', () => {
  beforeEach(() => {
    jest.clearAllMocks();
    el = shallow(<GradesView updateQueryParams={updateQueryParams} />);
  });
  describe('behavior', () => {
    it('initializes component hooks', () => {
      expect(useGradesViewData).toHaveBeenCalled();
    });
  });
  describe('render', () => {
    test('snapshot', () => {
      expect(el).toMatchSnapshot();
    });
    test('filterBadges load close behavior from hook', () => {
      expect(el.find(FilterBadges).props().handleClose).toEqual(
        hookProps.handleFilterBadgeClose,
      );
    });
  });
});
