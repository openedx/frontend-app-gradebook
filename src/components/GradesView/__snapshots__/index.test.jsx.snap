// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`GradesView component render snapshot 1`] = `
<Fragment>
  <SpinnerIcon />
  <InterventionsReport />
  <h3
    className="step-message-1"
  >
    filter-step-heading
  </h3>
  <div
    className="d-flex justify-content-between"
  >
    <FilterMenuToggle />
    <SearchControls />
  </div>
  <FilterBadges
    handleClose={[MockFunction hooks.handleFilterBadgeClose]}
  />
  <StatusAlerts />
  <h3>
    gradebook-step-heading
  </h3>
  <div
    className="d-flex justify-content-between align-items-center mb-2"
  >
    <ScoreViewInput />
    <BulkManagementControls />
  </div>
  <FilteredUsersLabel />
  <GradebookTable />
  <PageButtons />
  <p>
    * 
    test-masters-hint
  </p>
  <EditModal />
  <ImportSuccessToast />
</Fragment>
`;
